Added the implementation for the Paper Implementation Project #763 "Quantum Compression Algorithm for Symmetric States"

FipNad opened the pull request:
I added the implementation for the Paper Implementation Project #763 "Quantum Compression Algorithm for Symmetric States" in the algorithms folder. I have created a Python file consisting of all the necessary functions that provide the functionality described in the paper.

FipNad commented:
Hi @orsa-classiq! Okay, I will write a notebook will with the necessary info (some experiments and markdown formatting?.

FipNad commented:
Hi! @TaliCohn @orsa-classiq! Should I open another PR with the specific files?

FipNad commented:
Hello @orsa-classiq @TaliCohn! I have deleted the file from the /algorithms and I have added the files in the specific files under the community folder :)

orsa-classiq commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    def apply_matrix(matrix: CArray[CArray[CReal]], q: QArray[QBit], p1: CInt, p2: CInt, p3: CInt):
</pre><p>indeed it is currently cumbersome to do what you want. Until the syntax is going to improve, I suggest the following: define a function that accepts 3 qubits. Then you can just call it with the relevant indices.</p><p>Within the function use <code>bind</code> for the qubits</p><p> </p><p>Also - put the bind inside a <code>within-apply</code> operation </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/895/discussion/#comment-2042087746'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

FipNad replied:
<p>done that</p>

orsa-classiq commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #9.    quantum_program1 = synthesize(model1)
</pre><p>show the circuit using the <code>show</code> command (then everyone will be able to open the circuit you've created)  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/895/discussion/#comment-2042087756'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

FipNad commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<p>Yes, I modified it. I also modified the type in U_f function.</p>

orsa-classiq reviewed (changes requested):
Great work @FipNad !!
Please see my comments and fix accordingly.

FipNad commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<p>Done this</p>

FipNad replied:
<p>Also I added explanation for the U_f gate</p>

FipNad commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<p>done</p>

FipNad commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<p>done</p>

FipNad commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<p>done</p>

FipNad commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,533 @@
+{
```
<p>I thought that only one example needs to be uncommented. I uncommented the other two examples as well.</p>

FipNad commented:
@orsa-classiq I have modified the notebook based on your comments and sugestions.

orsa-classiq commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,571 @@
+{
```
<p>what is the complexity in terms of depth?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

orsa-classiq replied:
<p>I would assume that the user hasen't read the paper, and so I think that the unitaries U, U_ab and U_f are quite confusing names. Maybe add a diagram that explains their relations? because the high level structure is missing</p><p> maybe also find a bit better names for their function?</p><p> </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

orsa-classiq commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,571 @@
+{
```
<p>In the markdown - either use $U_{ab}$ if you want to mention the unitary, or use \'U_ab\` if it is the qfunc or a symbol you describe. Fix for all occurences</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

FipNad replied:
done

orsa-classiq commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,571 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    def apply_MultiCNOT(q: QArray[QBit], controls: CArray[CInt]):
</pre><p>I think you missed my last comment in the previous section.</p><p>Instead of define the function <code>apply_MultiCNOT</code>, I think you can do the following:</p><p><code>control((q[idx1] &amp;  q[idx2] &amp; ...), lambda: X(q[target])</code> </p><p>Anyway then you don't need the first part and all the bindings  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

FipNad replied:
for this part, I cannot apply the proposed method because the qubits that need to be used as controls differs for every circuit so i cannot point out to specific qubits prior of constructing the circuit.

orsa-classiq commented:
I've added a few more comments after reading your fixed version. Notice that you need to rebase your branch as you have conflicts

orsa-classiq commented on community/paper_implementation_project/quantum_compression_algorithm_for_symmetric_states/quantum_compression_algorithm_for_symmetric_states.ipynb:
```diff
@@ -0,0 +1,583 @@
+{
```
<p>1 - usually small $n$ is used for the number of qubits, and $N$ is for the number of states</p><p> </p><p>2 - V gate is not defined</p><p> </p><p>3 - can you give some intuition on what each single block does? if possible I would mention it here.</p><p> </p><p>4 - BTW - I think that what the last part is doing ("The final step of the algorithm will be to transform each of the state...&nbsp;into a new state so that in total just a number of&nbsp;... log&nbsp;qubits will be used") - is a transformation of unary decoding to binary encoding, not? see this: <a href="https://docs.classiq.io/latest/explore/algorithms/dqi/dqi_max_xorsat?h=dqi#step-2-encoding-conversions" target="_blank">https://docs.classiq.io/latest/explore/algorithms/dqi/dqi_max_xorsat/?h=dqi#step-2-encoding-conversions</a>. Might give intuition to somebody</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

orsa-classiq commented:
@FipNad great work! see my comments, let me know if something is not clear about the `control` comment.
Also - It seems that mistakenly you now have 83 files in you pull request (maybe you merged instead of rebase). Please fix it (If you struggle too much, you can open a new PR with just your files).
almost there!

FipNad commented:
> @FipNad great work! see my comments, let me know if something is not clear about the `control` comment. Also - It seems that mistakenly you now have 83 files in you pull request (maybe you merged instead of rebase). Please fix it (If you struggle too much, you can open a new PR with just your files). almost there!

Hello! Yes, somehow I managed to mess the branches!🙃

regarding the modifications, I will try to add some intuition on the gates. Gate V is defined in the comments on the main function implementation. I will change N to n.

regarding the multi qubit not gate, the issue that I have with your suggestion is that control (q[1] & q[2] &…) would imply to know the indexes prior. But because I have a generalized implementation, I cannot deal with this (as far as I know).

orsa-classiq commented:
> > @FipNad great work! see my comments, let me know if something is not clear about the `control` comment. Also - It seems that mistakenly you now have 83 files in you pull request (maybe you merged instead of rebase). Please fix it (If you struggle too much, you can open a new PR with just your files). almost there!
> 
> Hello! Yes, somehow I managed to mess the branches!🙃
> 
> regarding the modifications, I will try to add some intuition on the gates. Gate V is defined in the comments on the main function implementation. I will change N to n.
> 
> regarding the multi qubit not gate, the issue that I have with your suggestion is that control (q[1] & q[2] &…) would imply to know the indexes prior. But because I have a generalized implementation, I cannot deal with this (as far as I know).

you can do somthing like:
```
for i in range(...):
   for j in range(...):
       for target in range(...):
             control(q[i] & q[j], lambda: q[target])
```

FipNad commented:
> > > @FipNad great work! see my comments, let me know if something is not clear about the `control` comment. Also - It seems that mistakenly you now have 83 files in you pull request (maybe you merged instead of rebase). Please fix it (If you struggle too much, you can open a new PR with just your files). almost there!
> > 
> > 
> > Hello! Yes, somehow I managed to mess the branches!🙃
> > regarding the modifications, I will try to add some intuition on the gates. Gate V is defined in the comments on the main function implementation. I will change N to n.
> > regarding the multi qubit not gate, the issue that I have with your suggestion is that control (q[1] & q[2] &…) would imply to know the indexes prior. But because I have a generalized implementation, I cannot deal with this (as far as I know).
> 
> you can do somthing like:
> 
> ```
> for i in range(...):
>    for j in range(...):
>        for target in range(...):
>              control(q[i] & q[j], lambda: q[target])
> ```

So this method will be more efficient than the actual one?

FipNad commented:
Hi @orsa-classiq! I modified the notebook with additional data so that the reader can get some more intuition about the algorithm. Regarding the multi-CNOT function, i have tested and the current way is still the only one that works.

Regarding the number of files that are changed in the PR, now there are only 5, but I don't know if i fixed it.

FipNad commented:
> Hi @orsa-classiq! I modified the notebook with additional data so that the reader can get some more intuition about the algorithm. Regarding the multi-CNOT function, i have tested and the current way is still the only one that works.
> 
> Regarding the number of files that are changed in the PR, now there are only 5, but I don't know if i fixed it.

I also added the PR #951 If you can check.
